### DISCUSSION TITLE
What could cause the user's account page to be forbidden.

### DISCUSSION THREAD

**Discussion by weinzierl** [keycloak/keycloak#12894] [Q&A] [answered]
https://example.com/realms/myrealm/account
lets my users log in with their username and password. Instead of their username they see "Anonymous" in the top right corner. All subpages are "Forbidden, You do not have access rights to this request".  The users have manage-account and view-profile roles for the account client.

The issue exists for all users in a particular realm A, existing users and newly created ones. User in the Master realm or realm B can access their user page fine. I compared realm A and B and could not find a difference. 

What could cause the user page to be forbidden?

**Comment by cgeorgilakis** [comment]
This is Keycloak bug in account console. If User has not the role manage-account he can not view or manage anything in account console.
 
I have opened [this PR](https://github.com/keycloak/keycloak/pull/10568) for solving it. Although Keycloak team interest in related [github discussion](https://github.com/keycloak/keycloak/discussions/10450), [PR](https://github.com/keycloak/keycloak/pull/10568) has not been reviewed yet.

A temporary - of cource bad - solution could be to create a composite role containing all account roles in account-console client. With it, if a user has the role view-profile and not the role manage-account, he is able to view his account and when he tries to update his profile Keycloak informs him that this action is forbidden. Similar for all other roles. Another problem is that user with role manage-account-link cannot perform account linking - manage-account role is needed.

**Comment by weinzierl** [comment, marked as answer]
The [following answer](https://keycloak.discourse.group/t/forbidden-access-to-account-console-with-identity-provider-account-user/15861/2?u=weinzierl) gave me the crucial hint:

> Problem solved. I had an issue with export/import realm configurations between environments and the forbidden access was due to missing scopes com account-console client.

In "Clients" in the "account-console" client I added "roles" to "Assigned Default Client Scopes".

* Clients
* account-console
* Client Scopes
* Add "roles" to "Assigned Default Client Scopes"

In "Client Scopes" in the "roles" scope under "Mappers" I added a "client roles" mapper.

* Client Scopes
* Roles
* Mappers
* client roles
* Mapper Type: User Client Role
* Token Claim Name: resource_access.${client_id}.roles

![3B203569-8983-4E45-80CA-351C7EE4D5DE](https://user-images.githubusercontent.com/687096/178964509-b35c60bb-7f0d-4ec6-b69d-a3d9cf14bfc8.png)



With both of these changes I could access the user's account page.

**Reply by IgorKomissarov** [reply]
It works, the pages are accessible, but you still see "Anonymous" instead of a user name. To fix that you need to add all other "client scopes" to account-console client, not just "roles". As shown on the screenshot:
![image](https://github.com/keycloak/keycloak/assets/100859544/3b167b68-06c0-49c3-8260-be26ea635147)

**Comment by amdev74** [comment]
Hey, could someone help me, I have this Forbidden issue

`account-console` is well default assigned to everyone

If I add `resource_access.${client_id}.roles` it works

But this add roles from any client and it's not efficient for us and not possible

So I tried 

`resource_access.account-console.roles`

The output in the accesstoken /userinfo is exactly the same but I still have the forbidden access issue 

Can you give me the exact requirement for the console account ? what role should appear under which claim ? 

I think an error in the code beeing to specific on this check is possible

**Comment by amdev74** [comment]
This is an alternative version of the propsed answer

It fit more my use case for performence reason I can t just put client roles from any client in the scope "roles"

![Screenshot 2023-12-12 at 15 28 06](https://github.com/keycloak/keycloak/assets/12050544/c612a964-e074-4651-98c1-7d63dbafc4e8)

So it's very weird, same thing on account-console was not working, but resource_access.${client_id}.roles on account-console was working

Finally `resource_access.account.roles` was the solution, account-console is for console usage tho and account for the real account page on the UI. 

There is still a very weird behavior, if I create a clone of "roles" scope, exactly the same with the exact same `resource_access.account.roles` config as client role mapper. It doesn't works even if the Access token output is exactly the same. I think there is some weird condition in the code to access this page, it should just be having "manage-account" roles but it s not.

Also if it s not under the exact claim name `resource_access.account.roles` it doesn't works, I guess the code hard check this, which is probably not optimat ?
